services/QuillLMS/app/services/diagnostic_insights/skill_performance_by_classroom.rb

```diff
@@ -0,0 +1,191 @@
+# frozen_string_literal: true
+
+module DiagnosticInsights
+  class SkillPerformanceByClassroom < ApplicationService
+    class UnknownDiagnosticType < StandardError; end
+
+    STARTER_DIAGNOSTIC = 'starter'
+    INTERMEDIATE_DIAGNOSTIC = 'intermediate'
+    ADVANCED_DIAGNOSTIC = 'advanced'
+    ELL_STARTER_DIAGNOSTIC = 'ell_starter'
+    ELL_INTERMEDIATE_DIAGNOSTIC = 'ell_intermediate'
+    ELL_ADVANCED_DIAGNOSTIC = 'ell_advanced'
+
+    DIAGNOSTIC_TYPE_MAPPING = {
+      STARTER_DIAGNOSTIC => Activity::STARTER_PRE_DIAGNOSTIC_IDS + Activity::STARTER_POST_DIAGNOSTIC_IDS,
+      INTERMEDIATE_DIAGNOSTIC => Activity::INTERMEDIATE_PRE_DIAGNOSTIC_IDS + Activity::INTERMEDIATE_POST_DIAGNOSTIC_IDS,
+      ADVANCED_DIAGNOSTIC => Activity::ADVANCED_PRE_DIAGNOSTIC_IDS + Activity::ADVANCED_POST_DIAGNOSTIC_IDS,
+      ELL_STARTER_DIAGNOSTIC => Activity::ELL_STARTER_PRE_DIAGNOSTIC_IDS + Activity::ELL_STARTER_POST_DIAGNOSTIC_IDS,
+      ELL_INTERMEDIATE_DIAGNOSTIC => Activity::ELL_INTERMEDIATE_PRE_DIAGNOSTIC_IDS + Activity::ELL_INTERMEDIATE_POST_DIAGNOSTIC_IDS,
+      ELL_ADVANCED_DIAGNOSTIC => Activity::ELL_ADVANCED_PRE_DIAGNOSTIC_IDS + Activity::ELL_ADVANCED_POST_DIAGNOSTIC_IDS
+    }.freeze
+
+    FULL_PROFICIENCY = 'full_proficiency'
+    PARTIAL_PROFICIENCY = 'partial_proficiency'
+    NO_PROFICIENCY = 'no_proficiency'
+    NO_DATA = 'no_data'
+
+    attr_reader :classroom_id, :diagnostic_type, :start_date, :end_date
+
+    def initialize(classroom_id, diagnostic_type, start_date: nil, end_date: nil)
+      @classroom_id = classroom_id
+      @diagnostic_type = validate_diagnostic_type(diagnostic_type)
+      @start_date = start_date
+      @end_date = end_date
+    end
+
+    def run
+      students
+        .then { |payload| find_all_skill_performances(payload) }
```

Comment:
This `.then` syntax is very readable, I like it. One nit, is that the params all being `payload` makes is seem like all the inputs and outputs are the same shape, when they are not. Consider using more specific names for the params.